doc/developer/design/20230806_durable_catalog_state.md

```diff
@@ -0,0 +1,402 @@
+
+- `persist_catalog_state`:
+    1. Creates a new persist write handle.
+    2. Read the shard upper into memory.
+    3. Store the boot timestamp in memory.
+- `is_initialized`: Checks if upper is equal to the minimum timestamp.
+- `open`:
+    - If the catalog is not initialized:
+        1. Use the write handle to compare and append a batch of initial catalog objects.
+        2. Update the in memory upper to the new upper.
+    - If the catalog is initialized:
+        1. Open a persist read handle.
+        2. Read a snapshot of the catalog state into memory.
+        3. Run migrations in memory (including incrementing the epoch).
+        4. Use the write handle to compare and append a batch of migration changes into persist.
+        5. Update the in memory upper to the new upper.
+- `check_open`:
+    - If the catalog is not initialized then as long as we're the leader return `Ok`.
+    - If the catalog is initialized:
+        1. Open a persist read handle.
+        2. Read a snapshot of the catalog state into memory.
+        3. Run migrations in memory.
+            - None of the objects will be cached in memory.
+- `open_read_only`:
+    - If the catalog is not initialized then fail.
+    - If the catalog is initialized:
+        1. Open a persist read handle.
+        2. Read a snapshot of the catalog state into memory.
+- `boot_ts`: Return boot timestamp from memory.
+- `epoch`: This will be stored the same as any other catalog object so this can be treated as a
+  normal [read](#reads).
+
+#### Reads
+
+All reads (except for the deploy generation) will perform the follow steps:
+
+1. Confirm leadership.
+2. Return an object from the in memory cache.
+
+Reading the deploy generation will perform the following steps:
+
+- If the catalog is open, then read perform normal read.
+- If the catalog is not open:
+    - If the catalog is not initialized then return `None`.
+    - If the catalog is initialized:
+        1. Open a persist read handle.
+        2. Read the deploy generation from persist.
+
+Reading the deployment generation needs to work before the catalog has been opened and before
+migrations have been run.
+
+#### Writes
+
+- `transaction`: Use objects from in memory cache to initialize a new `Transaction`.
+- `commit_transaction`:
+    1. Use the write handle to compare and append the batch of changes into persist.
+    2. Update the in memory upper to the new upper.
+    3. Update in-memory object state.
+- `set_X`:
+    1. For each object create the following batch of changes:
+        - If the key of the object already exists in memory, add a retraction to the batch.
+        - Add an addition to the batch of the new value.
+    2. Use the write handle to compare and append the batch of changes into persist.
+    3. Update the in memory upper to the new upper.
+    4. Update in-memory object state.
+
+#### Misc
+
+- `confirm_leadership`: Check that the persist shard upper is equal to the upper cached in memory.
```

Comment:
This will use `fetch_recent_upper`, which says

```
 /// This requires fetching the latest state from consensus and is therefore a potentially
 /// expensive operation.
```

As a side note, I also just noticed it says
```
    /// Importantly, this operation is not
    /// linearized with other write operations.
```

So if I understand correctly, checking that the upper of a shard hasn't changed isn't enough to guarantee that someone else hasn't written to it. Is that accurate @danhhz? Is there a way to fetch an upper that's linearized with writes?

Reply:
> definitely open to adding a new method that does guarantee linearization (well, subject to crdb's usual linearization caveats of reads) and some tests for it, if that's necessary for this

That would work to implement this, but I think we might be to use the listen API for this instead. The listen API is linearizable right? `confirm_leadership` could be implemented by listening for new events and see if anyone else has written something. `is_initialized` can listen for events and see if there are more than 0 events.

Does that sound like it would work?

Reply:
the listen API doesn't have anything like this, unfortunately. it blocks (as in an await point) until there is new data or progress information to pass up

Reply:
Ah OK. For the differential catalog state I'm pretty sure that we'll either need a linearizable way to get the current upper OR a way to fetch all recent events from the Listen API. The `adapterd` process will need some linearizable way to retrieve all of the catalog changes that have finished before some operation has started. 

I think a linearizable way to get the current upper is probably the way to go, since it sounds pretty straightforward to add.